compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -325,6 +329,30 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         )
     }
 
+    /// If branch coverage is enabled, inject marker statements into each match
+    /// arm, and record their IDs in the table of branches.
+    ///
+    /// Unlike some of the other branch coverage visitor methods, this one needs
+    /// access to private details of [`Candidate`], so having it here is easier.
+    fn visit_coverage_match_expr(&mut self, candidates: &[&mut Candidate<'_, 'tcx>]) {
+        // Bail out if branch coverage is not enabled for this function.
+        let Some(branch_info) = self.coverage_branch_info.as_mut() else { return };
+
+        // FIXME(#124118) The current implementation of branch coverage for
+        // match arms can't handle or-patterns or guards.
+        if candidates.len() < 2
+            || candidates.iter().any(|c| c.pre_binding_block.is_none() || c.has_guard)
+        {
+            return;
+        }
```

Comment:
This is where we currently avoid adding branch coverage for match expressions that contain any or-patterns or guards.

(This is an improvement over previous drafts, which would ICE on or-patterns, and give incorrect counts for match expressions containing guards.)

For or-patterns, we can probably dig deeper into the pattern and find a suitable block, but I'll have to pay close attention to make sure that what I'm doing is correct.

For guards, we need to track enough information to let the instrumentor subtract the guard-was-false path from the sum of subsequent arms.